### PR TITLE
[RISCV] Convert AVLs with vlenb to VLMAX where possible

### PR DESCRIPTION
**Pull request by lukel97**
Given an AVL that's computed from vlenb, if it's equal to VLMAX then we can replace it with the VLMAX sentinel value.

The main motiviation is to be able to express an EVL of VLMAX in VP intrinsics whilst emitting vsetvli a0, zero, so that we can replace llvm.riscv.masked.strided.{load,store} with their VP counterparts.

This is done in RISCVVectorPeephole (previously RISCVFoldMasks, renamed to account for the fact that it no longer just folds masks) instead of SelectionDAG since there are multiple places places where VP nodes are lowered that would have need to have been handled.

This also avoids doing it in RISCVInsertVSETVLI as it's much harder to lookup the value of the AVL, and in RISCVVectorPeephole we can take advantage of DeadMachineInstrElim to remove any leftover PseudoReadVLENBs.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVFoldMasks.cpp [hunk 1]
```diff
@@ -62,6 +69,54 @@ char RISCVFoldMasks::ID = 0;
 
 INITIALIZE_PASS(RISCVFoldMasks, DEBUG_TYPE, "RISC-V Fold Masks", false, false)
 
+// If an AVL is a VLENB that's possibly scaled to be equal to VLMAX, convert it
+// to the VLMAX sentinel value.
+bool RISCVFoldMasks::convertToVLMAX(MachineInstr &MI) const {
+  if (!RISCVII::hasVLOp(MI.getDesc().TSFlags) ||
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:75]: Maybe we need a `isRVVPseudo` helper, I think it can be very useful.

**Comment by wangpc-pp** [comment]
I think we can rename `RISCVFoldMasks` to `PostISelRVVOptimization` or whatever.

File: llvm/lib/Target/RISCV/RISCVFoldMasks.cpp (hunk 1 above)
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:75]: Yeah, I think RISCVInsertVSETVLI relies on hasSEWOp to check if an instruction is a pseudo.

**Comment by lukel97** [comment]
> I think we can rename `RISCVFoldMasks` to `PostISelRVVOptimization` or whatever.

Maybe something like `RISCVVectorPeephole`?

**Comment by wangpc-pp** [comment]
> > I think we can rename `RISCVFoldMasks` to `PostISelRVVOptimization` or whatever.
> 
> Maybe something like `RISCVVectorPeephole`?

Make sense to me.

File: llvm/lib/Target/RISCV/RISCVFoldMasks.cpp (hunk 1 above)
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:75]: I think we could use `RISCVVPseudosTable::getPseudoInfo(Opc)` to check if it's a pseudo.

For here do we need to use it, or is it ok to just check for VL and SEW ops?

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:75]: Looking up table may be slow, I think it's OK to just check VL/SEW.

File: llvm/lib/Target/RISCV/RISCVFoldMasks.cpp [hunk 2]
```diff
@@ -62,6 +69,54 @@ char RISCVFoldMasks::ID = 0;
...
+  }
+
+  if (!Def || Def->getOpcode() != RISCV::PseudoReadVLENB)
+    return false;
+
+  auto LMUL = RISCVVType::decodeVLMUL(RISCVII::getLMul(MI.getDesc().TSFlags));
+  // Fixed-point value, denominator=8
+  unsigned LMULFixed = LMUL.second ? (8 / LMUL.first) : 8 * LMUL.first;
+  unsigned SEW =
+      1 << MI.getOperand(RISCVII::getSEWOpNum(MI.getDesc())).getImm();
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:103]: Does this need to account for mask instructions having an SEW operand of 0?

**Comment by lukel97** [comment]
Ping

File: llvm/lib/Target/RISCV/RISCVVectorPeephole.cpp [hunk 3]
```diff
@@ -1,19 +1,30 @@
-//===- RISCVFoldMasks.cpp - MI Vector Pseudo Mask Peepholes ---------------===//
+//===- RISCVVectorPeephole.cpp - MI Vector Pseudo Peepholes --------------===//
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVVectorPeephole.cpp:1]: Is this only 79 characters?

File: llvm/lib/Target/RISCV/RISCVVectorPeephole.cpp [hunk 4]
```diff
@@ -58,11 +70,62 @@ class RISCVFoldMasks : public MachineFunctionPass {
...
+    return false;
+  MachineInstr *Def = MRI->getVRegDef(VL.getReg());
+  if (!Def)
+    return false;
+
+  // Fixed-point value, denominator=8
+  unsigned ScaleFixed = 8;
+  // Check if the VLENB was potentially scaled with slli/srli
+  if (Def->getOpcode() == RISCV::SLLI) {
+    ScaleFixed <<= Def->getOperand(2).getImm();
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVVectorPeephole.cpp:95]: Do we need to check that the shift amount isn't out of bounds?

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**